lisp/bazel-build.el

```diff
@@ -28,35 +28,111 @@
+  (let ((stderr-file (make-temp-file "bazel-stderr")))
+    (unwind-protect
+        (string-rstrip
+         (with-output-to-string
+           (let ((status
+                  (apply #'call-process bazel-command
+                         nil (list standard-output stderr-file) nil args)))
+             (if (/= status 0)
+                 (with-temp-buffer
+                   (insert-file-contents stderr-file)
+                   (error "Error running command: %s" (buffer-string)))))))
+      (if (file-exists-p stderr-file)
+	  (delete-file stderr-file)))))
+
+(defun bazel-query (&rest args)
+  "Run a bazel query subcommand. Return stdout as string."
+  (apply #'bazel-build--call-process (append bazel-query-args args)))
+
+(defun bazel-build--target-for-file (filename)
+  "Get the list of targets which includes the given filename."
+  (let* ((default-directory (file-name-directory filename))
+         ;; Resolve label for file with bazel query.
+         (fullname (bazel-query (file-name-nondirectory filename))))
+    ;; Produce target the file-label is in using bazel query.
+    (let ((cmd (format "attr('srcs', %s, %s:*)" fullname
+                       (car (split-string fullname ":")))))
+      (split-string (bazel-query cmd)))))
+
+(defun bazel-build--target-for-directory (dirname)
+  "Get the list of targets under the given directory name."
+  (let* ((default-directory dirname)
+         (results (split-string
+                   (bazel-query "kind('.*rule', ':*')")))
+         (package (car (split-string (car results) ":"))))
+    (append (list (concat package ":all")) results)))
+
+(defun bazel-build--target-for-directory-or-filename (file-or-dir)
+  "Get the list of targets under the given file or directory name."
+  (if (file-directory-p file-or-dir)
+      (bazel-build--target-for-directory file-or-dir)
+    (bazel-build--target-for-file file-or-dir)))
+
+(defun bazel-build--read-target (prompt &optional filename)
+  "Read a target name for the given or current file or dired directory name."
+  ;; Bazel query invocation can be slow, issue a message.
+  (message "Generating completions...")
+  (let* ((targets (bazel-build--target-for-directory-or-filename
+                   (or
+                    ;; A given filename.
+                    filename
+                    ;; Open on a BUILD file.
+                    (let* ((file-name (buffer-file-name)))
+                      (when (and file-name
+                                 (string= (file-name-nondirectory file-name) "BUILD"))
+                        (directory-file-name (file-name-directory file-name))))
+                    ;; The buffer filename.
+                    (buffer-file-name)
+                    ;; Open on a dired directory.
+                    (and dired-directory (directory-file-name dired-directory))))))
+    (completing-read (format "Target for %s: " prompt) targets nil nil (car targets))))
+
+;; TODO(blais): There may be a cl equivalent for this.
```

Comment:
This looks like the same as string-trim-right from subr-x.

Reply:
I love it. Switched. 
(Will start using this dependency going forward; still not clear to me which are the most lightweight and official extension libraries for basic manipulation functions like that that really ought to be part of the standard library.)